### PR TITLE
Add ability to bind a shared service instance

### PR DESCRIPTION
**Pull request by Samze**
As an app dev (receiver), I can see service instances that have been shared with me [#150801602](https://www.pivotaltracker.com/n/projects/2105761/stories/150801602)

**NOTE**: This PR builds on top of #932 &  #940 which should be merged first. The actual changes ontop of #940 can be viewed in [this DIFF](https://github.com/cloudfoundry-incubator/cloud_controller_ng_sapi/compare/pr-service-instance-share-summary...cloudfoundry-incubator:pr-service-instance-sharing-bind).

## Why

The services api team have been working on service instance sharing feature as documented in these
[inception notes](https://docs.google.com/document/d/1BYLonYwsMLyjQG0FM8Q-qtoCIpkCz1OUX87odpNKl6Y/edit#heading=h.7tqm9trjzejv). 

This feature will allow space developers to share service instances into other spaces for binding. You can see a broad summary in [these slides](https://docs.google.com/presentation/d/1NMGK5dzW7WEqJ5bUjX6uXPRThb1v-oRthQFHaphpCLE/edit#slide=id.gd3f84e0d5_3_52).

[API design doc](https://docs.google.com/document/d/1V5m7L6TPMdR1phRWqzjzCoQIlw44FWZHjKxBsDjUaQQ/edit#)

## What

This PR allows creation of binds to shared service instances through the API. It allows for POSTing to the  `/v2/service_binding` endpoint in a space with a shared instance.

Changes:
* Apps are bindable if the service instance is in the same space OR shared into the space.

Notable things that are not in this PR but are addressed in future stories are:
1. Supporting all the API endpoints required for `cf bind-service`
1. APIs for retrieving bindings

Notes:
1. Additional CATs for this change can be found [here](https://github.com/cloudfoundry-incubator/cf-acceptance-tests-sapi/commit/3195d32d7d19c10b91b06bc5b92373a0cb402832).

While the changes in this PR include the ancestor commits of  #932 & #940 it only technically depends on #932. We have included both as that's how we have naturally been working on this work (our feature *master* branch). Let us know if this poses an issue.

## PR 

* [X] I have viewed signed and have submitted the Contributor License Agreement
* [X] I have made this pull request to the `master` branch
* [X] I have run all the unit tests using `bundle exec rake`
* [X] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats) on bosh lite

Sapi

### CODE REVIEW

File: spec/unit/models/services/service_binding_spec.rb [hunk 1]
```diff
@@ -75,13 +75,59 @@ module VCAP::CloudController
...
+                }.to raise_error(Sequel::ValidationFailed, /service_instance space_mismatch/)
+              end
+            end
+
+            context 'when the service instance has been shared into the app space' do
+              before do
+                service_instance.add_shared_space(app.space)
+              end
+
+              it 'it is valid' do
```
Comment (lisamcho) [inline spec/unit/models/services/service_binding_spec.rb]: apologies for nitpicking. stylistically can you remove duplicate `its` for consistency with rest of codebase?. --> it 'is valid' do ... end

File: app/models/services/service_binding.rb [hunk 2]
```diff
@@ -44,8 +44,13 @@ def validate
 
     def validate_space_match
       return unless service_instance && app
+      return if service_instance.space == app.space
 
-      unless service_instance.space == app.space
+      if FeatureFlag.enabled?(:service_instance_sharing)
```
Comment (lisamcho) [inline app/models/services/service_binding.rb]: could the doubly-nested if be refactored?

**Comment by lisamcho** [comment]
Hi @Samze,
We also noticed that there aren't any docs in this PR. This would actually return shared service bindings (if they existed). Except there's no way to create them yet. To clarify, is there a separate story for those docs?

**Comment by ablease** [comment]
Hello @lisamcho 

Thanks for the feedback! We removed the extra "it"s and replaced the double-nested `if` with a single line.

In your comment that "there's no way to create them yet," did you mean there's no way to create shared service instances, or bindings to shared service instances?

[This merged PR](https://github.com/cloudfoundry/cloud_controller_ng/pull/932) added the ability to "create" shared instances (i.e., share a service instance). Also in this PR we [added documentation](https://github.com/cloudfoundry/cloud_controller_ng/pull/932/files#diff-11604cbc3286fa6ba77443c8c299f9fe) in the v3 docs, which introduces the concept of shared service instances, and how to create them. 

If you meant the bindings themselves, this PR adds the ability to create bindings to shared service instances. We didn't see the need for API-level doc changes, since the create service binding API endpoint remains the same.

Also we realised that most of the commits had the wrong story id, also the link at the top of this PR links to the wrong story. The correct story is [#150801602](https://www.pivotaltracker.com/n/projects/2105761/stories/150801602). Sorry about that. 

Thanks again, 

Alex and @jenspinney

**Comment by lisamcho** [comment]
Thanks! I didn't see that docs had already been merged in.

**Pull request merged**